meson's function check code mis-detects sincosf on macOS using GCC-12

```
arm64-apple-darwin22-g++ -Isrc/libharfbuzz.0.dylib.p -Isrc -I../harfbuzz-7.3.0/ src -I. -I../harfbuzz-7.3.0 -I/Users/fabian/Gentoo-13/usr/include/freetype2 -I/ Users/fabian/Gentoo-13/usr/include/glib-2.0 -I/Users/fabian/Gentoo-13/usr/lib/  glib-2.0/include -fdiagnostics-color=always -Wall -Winvalid-pch -std=c++11 -fno-exceptions -fno-rtti -fno-exceptions -fno-rtti -fno-threadsafe-statics -        fvisibility-inlines-hidden -DHAVE_CONFIG_H -O2 -pipe -DHB_NDEBUG -Wno-non-      virtual-dtor -MD -MQ src/libharfbuzz.0.dylib.p/hb-ot-face.cc.o -MF src/         libharfbuzz.0.dylib.p/hb-ot-face.cc.o.d -o src/libharfbuzz.0.dylib.p/hb-ot-face.cc.o -c ../harfbuzz-7.3.0/src/hb-ot-face.cc
In file included from ../harfbuzz-7.3.0/src/OT/glyf/Glyph.hh:10,
                 from ../harfbuzz-7.3.0/src/OT/glyf/glyf.hh:13,
                 from ../harfbuzz-7.3.0/src/hb-ot-glyf-table.hh:33,
                 from ../harfbuzz-7.3.0/src/hb-ot-face.cc:30:
../harfbuzz-7.3.0/src/OT/glyf/VarCompositeGlyph.hh: In static member function   ‘static void OT::glyf_impl::VarCompositeGlyphRecord::rotate(float (&)[4], OT::  contour_point_t&, float)’:
../harfbuzz-7.3.0/src/OT/glyf/VarCompositeGlyph.hh:184:5: error: ‘sincosf’ was  not declared in this scope; did you mean ‘sinhf’?
  184 |     sincosf (rotation, &s, &c);
      |     ^~~~~~~
      |     sinhf
```

this is because HAVE_SINCOSF is defined:
```
Checking for function "sincosf" : YES 
```
which is incorrect, however it appears the buildsystem gets confused by the compiler built-in:
```
Running compile:
Working directory:  /Users/fabian/Gentoo-13/var/tmp/portage/media-libs/harfbuzz-7.3.0/work/harfbuzz-7.3.0-.arm64/meson-private/tmp_uyjf8tz
Command line:  arm64-apple-darwin22-g++ /Users/fabian/Gentoo-13/var/tmp/portage/media-libs/harfbuzz-7.3.0/work/harfbuzz-7.3.0-.arm64/meson-private/tmp_uyjf8tz/ testfile.cpp -o /Users/fabian/Gentoo-13/var/tmp/portage/media-libs/harfbuzz-7.3.0/work/harfbuzz-7.3.0-.arm64/meson-private/tmp_uyjf8tz/output.exe -O2 -pipe -   DHB_NDEBUG -O0 -fpermissive -std=c++11 -fno-exceptions -fno-rtti -Wl,-          dead_strip_dylibs

Code:

        #define sincosf meson_disable_define_of_sincosf

        #include <limits.h>
        #undef sincosf

        #ifdef __cplusplus
        extern "C"
        #endif
        char sincosf (void);

        #if defined __stub_sincosf || defined __stub___sincosf
        fail fail fail this function is not going to work
        #endif

        int main(void) {
          return sincosf ();
        }
Compiler stdout:

Compiler stderr:
 Undefined symbols for architecture arm64:
  "_sincosf", referenced from:
      _main in ccfMdc8T.o
ld: symbol(s) not found for architecture arm64
collect2: error: ld returned 1 exit status

Running compile:
Working directory:  /Users/fabian/Gentoo-13/var/tmp/portage/media-libs/harfbuzz-7.3.0/work/harfbuzz-7.3.0-.arm64/meson-private/tmpvpv_4xvt
Command line:  arm64-apple-darwin22-g++ /Users/fabian/Gentoo-13/var/tmp/portage/media-libs/harfbuzz-7.3.0/work/harfbuzz-7.3.0-.arm64/meson-private/tmpvpv_4xvt/ testfile.cpp -o /Users/fabian/Gentoo-13/var/tmp/portage/media-libs/harfbuzz-7.3.0/work/harfbuzz-7.3.0-.arm64/meson-private/tmpvpv_4xvt/output.exe -O2 -pipe -   DHB_NDEBUG -O0 -fpermissive -std=c++11 -fno-exceptions -fno-rtti -Wl,-          dead_strip_dylibs

Code:

        int main(void) {

        /* With some toolchains (MSYS2/mingw for example) the compiler
         * provides various builtins which are not really implemented and
         * fall back to the stdlib where they aren't provided and fail at
         * build/link time. In case the user provides a header, including
         * the header didn't lead to the function being defined, and the
         * function we are checking isn't a builtin itself we assume the
         * builtin is not functional and we just error out. */
        #if !1 && !defined(sincosf) && !0
            #error "No definition for __builtin_sincosf found in the prefix"
        #endif

        #ifdef __has_builtin
            #if !__has_builtin(__builtin_sincosf)
                #error "__builtin_sincosf not found"
            #endif
        #elif ! defined(sincosf)
            __builtin_sincosf;
        #endif
        return 0;
        }
Compiler stdout:

Compiler stderr:

Checking for function "sincosf" : YES
```
I suppose if the builtin is supposed to be used, then some macro or define needs to ne set so the code will actually use the builtin?

--- a/meson.build
+++ b/meson.build
@@ -72,15 +72,15 @@ check_headers = [
 ]
 
 check_funcs = [
-  ['atexit'],
-  ['mprotect'],
-  ['sysconf'],
-  ['getpagesize'],
-  ['mmap'],
-  ['isatty'],
-  ['uselocale'],
-  ['newlocale'],
-  ['sincosf'],
+  ['atexit', {'prefix': '#include <stdlib.h>'}],
+  ['mprotect', {'prefix': '#include <sys/mman.h>'}],
+  ['sysconf', {'prefix': '#include <unistd.h>'}],
+  ['getpagesize', {'prefix': '#include <unistd.h>'}],
+  ['mmap', {'prefix': '#include <sys/mman.h>'}],
+  ['isatty', {'prefix': '#include <unistd.h>'}],
+  ['uselocale', {'prefix': '#include <locale.h>'}],
+  ['newlocale', {'prefix': '#include <locale.h>'}],
+  ['sincosf', {'prefix': '#define _GNU_SOURCE\n#include <math.h>'}],
 ]
 
 m_dep = cpp.find_library('m', required: false)
@@ -356,11 +356,12 @@ foreach check : check_funcs
   opts = check.get(1, {})
   link_withs = opts.get('link_with', [])
   check_deps = opts.get('deps', [])
+  check_prefix = opts.get('prefix', '')
   extra_deps = []
   found = true
 
   # First try without linking
-  found = cpp.has_function(name, dependencies: check_deps)
+  found = cpp.has_function(name, prefix: check_prefix, dependencies: check_deps)
 
   if not found and link_withs.length() > 0
     found = true
@@ -375,7 +376,7 @@ foreach check : check_funcs
     endforeach
 
     if found
-      found = cpp.has_function(name, dependencies: check_deps + extra_deps)
+      found = cpp.has_function(name, prefix: check_prefix, dependencies: check_deps + extra_deps)
     endif
   endif
 

